src/transformers/generation/utils.py

```diff
@@ -1803,17 +1803,28 @@ def generate(
                 f"Passing both `cache_implementation` (used to initialize certain caches) and `{cache_name}` (a "
                 "Cache object) is unsupported. Please use only one of the two."
             )
+        elif not generation_config.use_cache:
+            pass  # no need to setup cache if use_cache=False
         elif generation_config.cache_implementation is not None:
             if generation_config.cache_implementation in NEED_SETUP_CACHE_CLASSES_MAPPING:
                 if generation_config.cache_implementation == "static" and not self._supports_static_cache:
                     raise ValueError(
                         "This model does not support `cache_implementation='static'`. Please check the following "
                         "issue: https://github.com/huggingface/transformers/issues/28981"
                     )
+                max_cache_length = generation_config.max_length
+                if (
+                    inputs_tensor.shape[1] != input_ids_length
+                    and model_input_name == "inputs_embeds"
+                    and not self.config.is_encoder_decoder
+                ):
+                    max_cache_length += inputs_tensor.shape[1]
                 model_kwargs[cache_name] = self._get_cache(
                     cache_implementation=generation_config.cache_implementation,
-                    batch_size=generation_config.num_beams * generation_config.num_return_sequences * batch_size,
-                    max_cache_len=generation_config.max_length,
+                    batch_size=max(
+                        generation_config.num_return_sequences * batch_size, generation_config.num_beams * batch_size
+                    ),
```

Comment:
```suggestion
                    batch_size=max(generation_config.num_return_sequences, generation_config.num_beams) * batch_size
```

Reply:
Yes, the thing is that if `num_beams` is defined along with `num_return_sequences` we should not be counting that together. Because the total batch size is always `beam * bs`. 

The `num_return_sequences` can be used with sampling, where the batch size will be `num_return_seq * bs`. I think getting the max works but we can also explicitly check in which generation strategy we are